File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/screens/main/viewmodel/DelayedQuery.java
package ru.usedesk.knowledgebase_gui.screens.main.viewmodel;

import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.NonNull;

import java.util.concurrent.TimeUnit;

import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.PublishSubject;

class DelayedQuery {
    private final PublishSubject<String> queryPublishSubject = PublishSubject.create();
    private final Disposable disposable;

    DelayedQuery(@NonNull final MutableLiveData<String> searchQueryLiveData, int delayMilliseconds) {
        disposable = queryPublishSubject.debounce(delayMilliseconds, TimeUnit.MILLISECONDS)
                .subscribe(searchQueryLiveData::postValue);
    }

    void dispose() {
        disposable.dispose();
    }

    void onNext(String query) {
        queryPublishSubject.onNext(query);
    }
}

File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/internal/screens/main/DelayedQuery.kt
package ru.usedesk.knowledgebase_gui.internal.screens.main

import androidx.lifecycle.MutableLiveData
import io.reactivex.disposables.Disposable
import io.reactivex.subjects.PublishSubject
import java.util.concurrent.TimeUnit

internal class DelayedQuery(
        searchQueryLiveData: MutableLiveData<String>,
        delayMilliseconds: Int
) {

    private val queryPublishSubject = PublishSubject.create<String>()
    private val disposable: Disposable = queryPublishSubject.debounce(
            delayMilliseconds.toLong(),
            TimeUnit.MILLISECONDS
    ).subscribe {
        searchQueryLiveData.postValue(it)
    }

    fun dispose() {
        disposable.dispose()
    }

    fun onNext(query: String) {
        queryPublishSubject.onNext(query)
    }
}